### ISSUE TITLE
High memory usage during server resync, server generally crashes after an hour

### ISSUE THREAD

**Issue by push-build-user** [couchbase/sync_gateway#1116]
We are running Couchbase server enterprise 3.0.2 and sync gateway enterprise 1.1.0 on Azure VM Ubuntu 12.04 with 4 2.0Ghz CPUs and 7 GB of RAM. We have about 2 million documents in our database and about 3000 active users, with maybe 100 concurrent.
We recently did a resync that added a lot of channels to sync gateway (one per week per user). The resync itself took about 8-12 hours and used a lot of memory but didn't crash. Afterwards sync gateway ran so slowly it basically made our whole production instance unusable. We had previously tested this configuration on our development server (far less users and data) and on a copy of our production server (same amount of data, but far less load on the server than real production) and on both of those it ran fine. On our production instance sync gateway keeps using more and more memory, until it eventually crashes the node and the VM resets.

**Comment by zgramana** [comment]
1.1.1 addresses a number of these issues directly, but there are some test scenarios this brings up that we need to capture.

**Comment by push-build-user** [comment]
By starting a new VM up and only running sync gateway on it I have been able to run the resync ok so far. It's been running for a few hours now and it's only using on average 10-15% of memory. The couchbase server node it is connected to though is showing over 80% memory usage. So the crashes could have been related to couchbase server. I upgraded one of our production VMs to have 14GB of memory, and now the sync gateway instance on that one is only using 15-20% of memory, while the sync gateway instance on the VM I left at 7GB is still using 40-50% of memory. Neither node has crashed while the resync has been running on the separate node, so I believe the crashes were because sync gateway was using a ton of memory querying the new views for all the new channels while running the resync. Or they could have been that it was using a lot of memory indexing the new channels in its views and now it has finished that process.
